src/compute/src/containers/stack.rs

```diff
@@ -0,0 +1,268 @@
+impl<T: Ord + Columnation + ToOwned<Owned = T> + 'static> BatchContainer for ChunkedStack<T> {
+    type PushItem = T;
+    type ReadItem<'a> = &'a Self::PushItem;
+
+    #[inline]
+    fn copy(&mut self, item: &T) {
+        self.copy(item);
+    }
+    #[inline]
+    fn copy_range(&mut self, other: &Self, start: usize, end: usize) {
+        let range = other.range(start..end);
+        self.reserve_items(range);
+        for item in range {
+            self.copy(item);
+        }
+    }
+
+    #[inline]
+    fn with_capacity(size: usize) -> Self {
+        Self::with_capacity(size)
+    }
+
+    fn merge_capacity(cont1: &Self, cont2: &Self) -> Self {
+        let mut new = Self::with_capacity(cont1.length + cont2.length);
+        new.reserve_regions(std::iter::once(cont1).chain(std::iter::once(cont2)));
+        new
+    }
+
+    #[inline]
+    fn index(&self, index: usize) -> Self::ReadItem<'_> {
+        self.index(index)
+    }
+
+    #[inline]
+    fn len(&self) -> usize {
+        self.len()
+    }
+}
+
+impl<T: Columnation> Index<usize> for ChunkedStack<T> {
+    type Output = T;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        ChunkedStack::index(self, index)
+    }
+}
+
+impl<T: Columnation> Default for ChunkedStack<T> {
+    fn default() -> Self {
+        Self::with_capacity(0)
+    }
+}
+
+impl<A: Columnation, B: Columnation, C: Columnation> ChunkedStack<(A, B, C)> {
+    /// Copies a destructured tuple `(A, B, C)` into this column stack.
+    ///
+    /// This serves situations where a tuple should be constructed from its constituents but
+    /// not all elements are available as owned data.
+    ///
+    /// The element can be read by indexing
+    pub fn copy_destructured(&mut self, r0: &A, r1: &B, r2: &C) {
+        let copy = unsafe { self.inner.copy_destructured(r0, r1, r2) };
+        self.push(copy);
+    }
+}
+
+/// An iterator of a [`ChunkedStack`].
+pub struct Iter<'a, T: Columnation> {
+    stack: &'a ChunkedStack<T>,
+    offset: usize,
+    limit: usize,
+}
+
+impl<'a, T: Columnation> Clone for Iter<'a, T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a, T: Columnation> Copy for Iter<'a, T> {}
+
+impl<'a, T: Columnation> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.limit {
+            None
+        } else {
+            let next = self.stack.index(self.offset);
+            self.offset = self.offset.saturating_add(1);
+            Some(next)
+        }
+    }
+}
+
+impl<T: Columnation> Drop for ChunkedStack<T> {
+    fn drop(&mut self) {
+        for array in &mut self.local {
+            // SAFETY: All elements in `array` have their allocations in a region. We drop the
+            // region and forget the immediate values.
+            unsafe {
+                array.set_len(0);
+            }
```

Comment:
I have a hard time following this part. Is it also important to avoid double frees? If we don't call `set_len(0)` here, we would drop parts of `inner` and then we would drop them again when we call clear/drop `inner`?

Reply:
Added a comment.